AKS Secure Baseline - Adding RBAC path (#204)
mspnp/aks-secure-baseline: 03-aad.md

# Prep for Azure Active Directory Integration

In the prior step, you [generated the user-facing TLS certificate](./02-ca-certificates.md); now we'll prepare Azure AD for Kubernetes role-based access control (RBAC). This will ensure you have an Azure AD security group(s) and user(s) assigned for group-based Kubernetes control plane access.

## Expected results

Following the steps below you will result in an Azure AD configuration that will be used for Kubernetes control plane (Cluster API) authorization.

> | Object                             | Purpose                                                 |

Reviewer: Can you check line endings on this file, I think they all changed.  Makes it harder to find out what specifically changed.  I'll review this file again once you revert that change.

   ```bash
   export TENANTDOMAIN_K8SRBAC=$(az ad signed-in-user show --query 'userPrincipalName' -o tsv | cut -d '@' -f 2 | sed 's/\"//')
   export AADOBJECTNAME_USER_CLUSTERADMIN=bu0001a000800-admin
   export AADOBJECTID_USER_CLUSTERADMIN=$(az ad user create --display-name=${AADOBJECTNAME_USER_CLUSTERADMIN} --user-principal-name ${AADOBJECTNAME_USER_CLUSTERADMIN}@${TENANTDOMAIN_K8SRBAC} --force-change-password-next-login --password ChangeMebu0001a0008AdminChangeMe --query objectId -o tsv)
   ```

1. Add the cluster admin user(s) to the cluster admin security group.

   > :book: The recently created break-glass admin user is added to the Kubernetes Cluster Admin group from Azure AD. After this step the Azure AD Admin team will have finished the app team's request.

   ```bash
   az ad group member add -g $AADOBJECTID_GROUP_CLUSTERADMIN --member-id $AADOBJECTID_USER_CLUSTERADMIN
   ```

1. Create/identify the Azure AD security group that is going to be a namespace reader.

   ```bash

>    export AADOBJECTID_GROUP_A0005_READER=$(az ad group create --display-name 'cluster-ns-a0005-readers-bu0001a000800' --mail-nickname 'cluster-ns-a0005-readers-bu0001a000800' --description "Principals in this group are readers of namespace a0005 in the bu0001a000800 cluster." --query objectId -o tsv)

Reviewer: Why is this called "A0005" when the namespace is `a0008`?  Looks like you have it as "0005" in a couple of places.  Can you update them all?  Thanks.

Author: Renamed !

   ```

## Kubernetes RBAC backing store

> In the prerequisites step you created a new tenant if you were not part of the User Administrator group in the tenant associated to your Azure subscription, if that is the case you will still leverage integrated authentication between Azure Active Directory (Azure AD) and AKS. When enabled, this integration allows customers to use Azure AD users, groups, or service principals as subjects in Kubernetes RBAC. This feature frees you from having to separately manage user identities and credentials for Kubernetes. However, you still have to set up and manage Azure.
> 
> On the other hand, if you are using a single tenant (the one associated with your Azure subscription) and you have enough privileges to create the AKS admin Active Directory Security Group, the cluster deployment step will take care of the Azure AD group association by making use of Azure RBAC for Kubernetes Authorization; this approach allows for the unified management and access control across Azure Resources, AKS, and Kubernetes resources.
> 
> 1. Set up groups to map into other Kubernetes Roles. _Optional, fork required._
> 
>    > :book:  Follow this step if you forked the repository and you have two separate tenants, since in that case you are making use of integrated authentication between Azure Active Directory (Azure AD) and AKS.
>    The team knows there will be more than just cluster admins that need group-managed access to the cluster. Out of the box, Kubernetes has other roles like _admin_, _edit_, and _view_ which can also be mapped to Azure AD Groups for use both at namespace and at the cluster level.
> 
>    In the [`cluster-rbac.yaml` file](./cluster-manifests/cluster-rbac.yaml) and the various namespaced [`rbac.yaml files`](./cluster-manifests/cluster-baseline-settings/rbac.yaml), you can uncomment what you wish and replace the `<replace-with-an-aad-group-object-id...>` placeholders with corresponding new or existing AD groups that map to their purpose for this cluster or namespace. You do not need to perform this action for this walk through; they are only here for your reference.
> 
>    :bulb: If your tenants are the same, you will allow Azure RBAC to be the backing store for your Kubernetes RBAC, so you will transparently make these group associations to [Azure RBAC roles](https://docs.microsoft.com/azure/aks/manage-azure-rbac) in the cluster deployment step. Above in the steps you created the Azure AD security group that is going to be a namespace reader, that Group ID will be associated to the 'Azure Kubernetes Service RBAC Reader' Azure RBAC built in role.

Reviewer: This is looking nice.  I spent a little time wordsmithing this a bit more.  What do you think about the following?

```suggestion
AKS supports backing Kubernetes with Azure AD in two different modalities. One is direct association between Azure AD and Kubernetes `ClusterRoleBindings`/`RoleBindings` in the cluster. This is possible no matter if the Azure AD tenant you wish to use to back your Kubernetes RBAC is the same or different than the Tenant backing your Azure resources. If however the tenant that is backing your Azure resources (Azure RBAC source) is the same tenant you plan on using to back your Kubernetes RBAC, then instead you can add a layer of indirection between Azure AD and your cluster by using Azure RBAC instead of direct cluster `RoleBinding` manipulation.  When performing this walkthrough, you may have had no choice but to associate the cluster with another tenant (due to the elevated permissions necessary in Azure AD to manage groups and users); but when you take this to production be sure you're using Azure RBAC as your Kubernetes RBAC backing store if the tenants are the same. Both cases still leverage integrated authentication between Azure AD and AKS, Azure RBAC simply elevates this control to Azure RBAC instead of direct yaml-based management within the cluster which usually will align better with your organization's governance strategy.

### Azure RBAC _[Preferred]_

If you are using a single tenant for this walkthrough, the cluster deployment step later will take care of the necessary role assignments for the groups created above.  Specifically, in the above steps, you created the Azure AD security group `cluster-ns-a0008-readers-bu0001a000800` that is going to be a namespace reader in namespace `a0008` and the Azure AD security group `cluster-admins-bu0001a000800` is is going to contain cluster admins. Those group Object IDs will be associated to the 'Azure Kubernetes Service RBAC Reader' and 'Azure Kubernetes Service RBAC Cluster Admin' RBAC role respectively, scoped to their proper level within the cluster.

Using Azure RBAC as your authorization approach is ultimately preferred as it allows for the unified management and access control across Azure Resources, AKS, and Kubernetes resources.  At the time of this writing there are four [Azure RBAC roles](https://docs.microsoft.com/azure/aks/manage-azure-rbac#create-role-assignments-for-users-to-access-cluster) that represent typical cluster access patterns.

### Direct Kubernetes RBAC management _[Alternative]_

If you instead wish to not use Azure RBAC as your Kubernetes RBAC authorization mechanism, either due to the intentional use of disparate Azure AD tenants or another business justifications, you can then manage these RBAC assignments via direct `ClusterRoleBinding`/`RoleBinding` associations. This method is also useful when the four Azure RBAC roles are not granular enough for your desired permission model.

1. Set up additional Kubernetes RBAC associations. _Optional, fork required._

   > :book:  The team knows there will be more than just cluster admins that need group-managed access to the cluster. Out of the box, Kubernetes has other roles like _admin_, _edit_, and _view_ which can also be mapped to Azure AD Groups for use both at namespace and at the cluster level. Likewise custom roles can be created which need to be mapped to Azure AD Groups.

   In the [`cluster-rbac.yaml` file](./cluster-manifests/cluster-rbac.yaml) and the various namespaced [`rbac.yaml files`](./cluster-manifests/cluster-baseline-settings/rbac.yaml), you can uncomment what you wish and replace the `<replace-with-an-aad-group-object-id...>` placeholders with corresponding new or existing Azure AD groups that map to their purpose for this cluster or namespace. **You do not need to perform this action for this walkthrough**; they are only here for your reference.
```

# Prep for Azure Active Directory Integration

In the prior step, you [generated the user-facing TLS certificate](./02-ca-certificates.md); now we'll prepare Azure AD for Kubernetes role-based access control (RBAC). This will ensure you have an Azure AD security group(s) and user(s) assigned for group-based Kubernetes control plane access.

## Expected results

Following the steps below you will result in an Azure AD configuration that will be used for Kubernetes control plane (Cluster API) authorization.

| Object                             | Purpose                                                 |
|------------------------------------|---------------------------------------------------------|
| A Cluster Admin Security Group     | Will be mapped to `cluster-admin` Kubernetes role.      |
| A Cluster Admin User               | Represents at least one break-glass cluster admin user. |
| Cluster Admin Group Membership     | Association between the Cluster Admin User(s) and the Cluster Admin Security Group. |
| A Namespace Reader Security Group  | Represents users that will have read-only access to a specific namespace in the cluster. |

> 

Reviewer: Looks like extra new line got added.

```suggestion
```

| _Additional Security Groups_       | _Optional._ A security group (and its memberships) for the other built-in and custom Kubernetes roles you plan on using. |

> 

Reviewer: Looks like extra new line got added.

```suggestion
```

## Steps

> :book: The Contoso Bicycle Azure AD team requires all admin access to AKS clusters be security-group based. This applies to the new AKS cluster that is being built for Application ID a0008 under the BU0001 business unit. Kubernetes RBAC will be AAD-backed and access granted based on users' AAD group membership(s).

> 

Reviewer: Looks like extra new line got added.

```suggestion
```
